### PR TITLE
feat: add awakened and focus bar

### PR DESCRIPTION
**Pull request by byBackfish**
Adds awakened mask bar (shaman ability, you need to spend mana using certain masks to activate it) 

image

Adds focus bar (archer ability, hitting a mob gives focus, missing a shot resets it)
image

### CODE REVIEW

**Comment by kristofbolyai** [comment]
Is awakened only possbile if there is no blood pool (so they don't conflict)?

**Comment by kristofbolyai** [comment]
Also, can we render awakened as red text?

**Comment by kristofbolyai** [comment]
Can we color focus bar differently? Just like powder overlay is colored.

**Comment by kristofbolyai** [comment]
> Can we color focus bar differently? Just like powder overlay is colored.

Also, if awakened is not related to HP, color that differently as well.

**Comment by kristofbolyai** [comment]
> > Can we color focus bar differently? Just like powder overlay is colored.
> 
> Also, if awakened is not related to HP, color that differently as well.

If you want, you can merge powder overlay into custom bars feature, and have a common base class for all of these.

**Comment by byBackfish** [comment]
> Also, can we render awakened as red text?

Wynncraft colors it white, so I thought white would be the best color to use for this

**Comment by kristofbolyai** [comment]
> > Also, can we render awakened as red text?
> 
> Wynncraft colors it white, so I thought white would be the best color to use for this

Then, make the bar white?

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 1]
```diff
@@ -183,6 +209,67 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
         }
     }
 
+    public static class AwakenedProgressBarOverlay extends HealthBarOverlay {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: No reason to inherit HealthBarOverlay. Now, it just messed up configs (textColor), and added an unused method for the base class. Make a new base class and use that.

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 2]
```diff
@@ -183,6 +209,67 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
+
+            String text = "Awakened: " + awakenedBar.current() + "/200";
+            renderText(poseStack, renderY, text);
+
+            float progress = (this.flip ? -1 : 1) * awakenedBar.progress();
+            renderColoredBar(poseStack, renderY, this.textColor, progress);
+        }
+    }
+
+    public static class FocusBarOverlay extends HealthBarOverlay {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: Same here obv.

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 3]
```diff
@@ -139,6 +146,25 @@ protected void renderBar(PoseStack poseStack, float renderY, float progress) {
                     progress);
         }
 
+        public void renderColoredBar(PoseStack poseStack, float renderY, CustomColor color, float progress) {
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: What is the reasoning for introducing `renderColoredBar` when `renderBar` already exists?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: It renders a bar which can be colored. There is no point in having it in the HealthBarOverlay class though. See my comments.

**Comment by Incompleteusern** [comment]
Aren't there a few more bars for other archetypes, not that they need to be implemented now?

**Comment by kristofbolyai** [comment]
> Aren't there a few more bars for other archetypes, not that they need to be implemented now?

I already added mana bank and blood bank.

**Comment by Incompleteusern** [comment]
> > Aren't there a few more bars for other archetypes, not that they need to be implemented now?
> 
> I already added mana bank and blood bank.

Doesn't warrior have some corrupted thing?

**Comment by byBackfish** [comment]
> > > Aren't there a few more bars for other archetypes, not that they need to be implemented now?
> > 
> > 
> > I already added mana bank and blood bank.
> 
> Doesn't warrior have some corrupted thing?

corrupted bar is displayed as a title with weird formatting, i sadly do not have a corrupted warrior right now tho

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java (hunk 3 above)
Comment (byBackfish) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: `renderBar` runs` RenderUtils.drawProgressBar`, whilst `renderColoredBar` runs `RenderUtils.drawColoredProgressBar`

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 4]
```diff
@@ -183,6 +189,172 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
         }
     }
 
+    public static class AwakenedProgressBarOverlay extends Overlay {
+        @Config(key = "feature.wynntils.customBarsOverlay.overlay.awakenedProgressBar.textShadow")
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: You only need to specify the key in special cases (where the auto generated one is not good, so when using inheritance). Remove them this case, only leave `@Config`

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 5]
```diff
@@ -183,6 +189,172 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
+                            renderY,
+                            0,
+                            this.textColor,
+                            this.getRenderHorizontalAlignment(),
+                            this.textShadow);
+        }
+    }
+
+    public static class FocusBarOverlay extends Overlay {
+        @Config(key = "feature.wynntils.customBarsOverlay.overlay.focusBar.textShadow")
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: Same with these

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 6]
```diff
@@ -183,6 +189,172 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
+            final float renderedHeight = universalBarTexture.height() / 2f * (this.getWidth() / 81);
+
+            float renderY =
+                    switch (this.getRenderVerticalAlignment()) {
+                        case Top -> this.getRenderY();
+                        case Middle -> this.getRenderY() + (this.getHeight() - renderedHeight) / 2;
+                        case Bottom -> this.getRenderY() + this.getHeight() - renderedHeight;
+                    };
+
+            String text = "Awakened: " + awakenedBar.current() + "/200";
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: I would kinda like the max value to be parsed from the boss bar like everything else does.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: Also this text is pretty large for the overlay, although this is not a breaking issue.

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java (hunk 4 above)
Comment (byBackfish) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: Would I still need to add it to the language file? Or is the name just the name of the variable?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: Basically the key you defined here is already the default one. So yes, keep the translations, but no need to set the key here.

Comment (byBackfish) [reply common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]: ah alright, thanks for telling me

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 7]
```diff
@@ -183,6 +189,172 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
+            renderText(poseStack, renderY, text);
+
+            float progress = (this.flip ? -1 : 1) * awakenedBar.progress();
+            renderColoredBar(poseStack, renderY, this.textColor, progress);
+        }
+
+        @Override
+        protected void onConfigUpdate(ConfigHolder configHolder) {}
+
+        public void renderColoredBar(PoseStack poseStack, float renderY, CustomColor color, float progress) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]:
```suggestion
        protected void renderColoredBar(PoseStack poseStack, float renderY, CustomColor color, float progress) {
```

File: common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java [hunk 8]
```diff
@@ -183,6 +189,172 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
...
+            renderText(poseStack, renderY, text);
+
+            float progress = (this.flip ? -1 : 1) * focus.progress();
+            renderColoredBar(poseStack, renderY, this.textColor, progress);
+        }
+
+        @Override
+        protected void onConfigUpdate(ConfigHolder configHolder) {}
+
+        public void renderColoredBar(PoseStack poseStack, float renderY, CustomColor color, float progress) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/overlays/CustomBarsOverlayFeature.java]:
```suggestion
        protected void renderColoredBar(PoseStack poseStack, float renderY, CustomColor color, float progress) {
```

**Review by kristofbolyai — DISMISSED** [review]

**Comment by kristofbolyai** [comment]
Looks good now! Thanks.

**Pull request merged**